In [54]:
import pandas as pd 

In [55]:
df = pd.read_csv("fifa21 raw data v2.csv", low_memory = False)
df.head(2)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [56]:
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

## Drop 'photoUrl' column since playerUrl will contain such information.

In [57]:
df.drop("photoUrl", axis = 1, inplace = True)

In [58]:
df.head(2)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [59]:
df.shape

(18979, 76)

## Capitalize the first letter of the 'playerUrl' column

In [60]:
df.rename(columns={"playerUrl":"PlayerUrl"}, inplace = True)

In [61]:
df.head(2)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Change column 'LongName' to 'FullName' and '↓OVA' to OverallRating

In [62]:
df.rename(columns={"LongName":"FullName"}, inplace = True)

In [63]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [64]:
df.rename(columns={"↓OVA":"OverallRating"}, inplace = True)

## Examine df['Loan Date End'] to know how many null value it has

In [65]:
#df['Loan Date End'].isnull().sum()

## Expand the columns to access all columns for examination and cleaning

In [66]:
pd.set_option('display.max_columns', None)

In [67]:
#pd.set_option('display.max_rows')

In [68]:
df.head()

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   

   OverallRating  POT                         Club     Contract   Positions  \
0             93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF   
1             92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW   
2             91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK   
3             91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM   
4             91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM   

  Height Weight Preferred Foot  BOV Best Position        Joined Loan Date End  \
0  170cm   72kg           Left   93            RW   Jul 1, 2004           NaN   
1  187cm   83kg          Right   92            ST  Jul 10, 2018           NaN   
2  188cm   87kg          Right   91            GK  Jul 16, 2014           NaN   
3  181cm   70kg          Right   91           CAM  Aug 30, 2015           NaN   
4  175cm   68kg          Right   91            LW   Aug 3, 2017           NaN   

     Value   Wage Release Clause  Attacking  Crossing  Finishing  \
0  €103.5M  €560K        €138.4M        429        85         95   
1     €63M  €220K         €75.9M        437        84         95   
2    €120M  €125K        €159.4M         95        13         11   
3    €129M  €370K          €161M        407        94         82   
4    €132M  €270K        €166.5M        408        85         87   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                70             91       88    470         96     93   
1                90             82       86    414         88     81   
2                15             43       13    109         12     13   
3                55             94       82    441         88     85   
4                62             87       87    448         95     88   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           94            91            96       451            91   
1           76            77            92       431            87   
2           14            40            30       307            43   
3           83            93            92       398            77   
4           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \


## Remove the '\n\n\n\n' in the club column

In [69]:
df.Club

0               \n\n\n\nFC Barcelona
1                   \n\n\n\nJuventus
2            \n\n\n\nAtlético Madrid
3            \n\n\n\nManchester City
4        \n\n\n\nParis Saint-Germain
                    ...             
18974             \n\n\n\nWuhan Zall
18975        \n\n\n\nOldham Athletic
18976             \n\n\n\nDerry City
18977       \n\n\n\nDalian YiFang FC
18978       \n\n\n\nDalian YiFang FC
Name: Club, Length: 18979, dtype: object

In [70]:
df.Club = df['Club'].str[4:].replace()

In [71]:
df.Club

0               FC Barcelona
1                   Juventus
2            Atlético Madrid
3            Manchester City
4        Paris Saint-Germain
                ...         
18974             Wuhan Zall
18975        Oldham Athletic
18976             Derry City
18977       Dalian YiFang FC
18978       Dalian YiFang FC
Name: Club, Length: 18979, dtype: object

## Examine the dataframe again

In [72]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1             92   92      Juventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## With df['Loan Date End'] having 17966 null values out of the 18979 values of the dataframe, it is obvious the column is less informative and important so we drop such unnecessary column to make the data frame neater

In [73]:
df.drop("Loan Date End", axis = 1, inplace = True)

In [74]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1             92   92      Juventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Change the '~' in df['Contract'] to '-' which is the standard way to seperate

In [75]:
df['Contract'] = df['Contract'].str.replace("~", "-")

In [76]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 - 2021  RW, ST, CF  170cm   72kg   
1             92   92      Juventus  2018 - 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Also in df['Positions'], the seperator is a comma, we have to change it to a pipe to make the df clean

In [77]:
df.Positions = df['Positions'].str.replace(",", "|")

In [78]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 - 2021  RW| ST| CF  170cm   72kg   
1             92   92      Juventus  2018 - 2022      ST| LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Examine the columns that still have null values

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 75 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   FullName          18979 non-null  object
 3   PlayerUrl         18979 non-null  object
 4   Nationality       18979 non-null  object
 5   Age               18979 non-null  int64 
 6   OverallRating     18979 non-null  int64 
 7   POT               18979 non-null  int64 
 8   Club              18979 non-null  object
 9   Contract          18979 non-null  object
 10  Positions         18979 non-null  object
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  Preferred Foot    18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  Best Position     18979 non-null  object
 16  Joined            18979 non-null  object
 17  Value       

## From the infomation gotten from the .info() above, we can see that the Hits still have about 2595 null values because 16384 - 18979 = 2595.

In [80]:
df.Hits.isnull().sum()

2595

## Get the missen Hits values

## Since the Hits column entails the number of times the players  profile has been viewed on the website, with the column being set to NaN, we have the assume it means their profile has not been viewed on the site, hence we set the value to 0

In [81]:
df.Hits = df.Hits.fillna('0')

In [82]:
df.Hits.isnull().sum()

0

In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 75 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   FullName          18979 non-null  object
 3   PlayerUrl         18979 non-null  object
 4   Nationality       18979 non-null  object
 5   Age               18979 non-null  int64 
 6   OverallRating     18979 non-null  int64 
 7   POT               18979 non-null  int64 
 8   Club              18979 non-null  object
 9   Contract          18979 non-null  object
 10  Positions         18979 non-null  object
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  Preferred Foot    18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  Best Position     18979 non-null  object
 16  Joined            18979 non-null  object
 17  Value       

In [84]:
#After changeing the null values in Hits to 0, using .info(), we can see that there are no longer any null values and all we have to do is know if there will be any other futher cleaning to do

## some values in the Hit columns contain 'K', so we change those values to thousand

In [85]:
result = df[df['Hits'].str.contains('K')]


In [86]:
result

ID             Name                       FullName  \
8      231747        K. Mbappé                  Kylian Mbappé   
29     233049        J. Sancho                   Jadon Sancho   
34     212198  Bruno Fernandes  Bruno Miguel Borges Fernandes   
68     231677      M. Rashford                Marcus Rashford   
105    239085       E. Haaland                 Erling Haaland   
106    212462      Alex Telles            Alex Nicolao Telles   
147    229880   A. Wan-Bissaka              Aaron Wan-Bissaka   
416    220651         Angeliño     José Ángel Esmoris Tasende   
544    229558     D. Upamecano                Dayot Upamecano   
687    255475           Antony      Antony Matheus dos Santos   
690    244778          Trincão  Francisco Mota Castro Trincão   
1011   246147     M. Greenwood                Mason Greenwood   
1265   253004        Ansu Fati                 Anssumane Fati   
1380   248243     E. Camavinga              Eduardo Camavinga   
1818   251804          S. Dest                   Sergiño Dest   
2579   231485      A. Tuanzebe                  Axel Tuanzebe   
2757   245371        T. Almada                  Thiago Almada   
3791   251854            Pedri           Pedro González López   
3893   205343    J. Akpa Akpro         Jean-Daniel Akpa Akpro   
4091   254796       N. Madueke                   Noni Madueke   
4096   253283     F. Pellistri              Facundo Pellistri   
4800   251470  C. De Ketelaere           Charles De Ketelaere   
5266   245541         G. Reyna                 Giovanni Reyna   
6159   251852       K. Adeyemi                  Karim Adeyemi   
6235   252371    J. Bellingham                Jude Bellingham   
8116   251570        R. Cherki                   Rayan Cherki   
8401   238071      D. Sterling                 Dujon Sterling   
15819  255611       D. Maldini                 Daniel Maldini   

                                               PlayerUrl    Nationality  Age  \
8      http://sofifa.com/player/231747/kylian-mbappe/...         France   21   
29     http://sofifa.com/player/233049/jadon-sancho/2...        England   20   
34     http://sofifa.com/player/212198/bruno-miguel-b...       Portugal   25   
68     http://sofifa.com/player/231677/marcus-rashfor...        England   22   
105    http://sofifa.com/player/239085/erling-haaland...         Norway   19   
106    http://sofifa.com/player/212462/alex-nicolao-t...         Brazil   27   
147    http://sofifa.com/player/229880/aaron-wan-biss...        England   22   
416    http://sofifa.com/player/220651/jose-angel-esm...          Spain   23   
544    http://sofifa.com/player/229558/dayot-upamecan...         France   21   
687    http://sofifa.com/player/255475/antony-matheus...         Brazil   20   
690    http://sofifa.com/player/244778/francisco-mota...       Portugal   20   
1011   http://sofifa.com/player/246147/mason-greenwoo...        England   18   
1265   http://sofifa.com/player/253004/anssumane-fati...          Spain   17   
1380   http://sofifa.com/player/248243/eduardo-camavi...         France   17   
1818   http://sofifa.com/player/251804/sergino-dest/2...  United States   19   
2579   http://sofifa.com/player/231485/axel-tuanzebe/...        England   22   
2757   http://sofifa.com/player/245371/thiago-almada/...      Argentina   19   
3791   http://sofifa.com/player/251854/pedro-gonzalez...          Spain   17   
3893   http://sofifa.com/player/205343/jean-daniel-ak...    Ivory Coast   27   
4091   http://sofifa.com/player/254796/noni-madueke/2...        England   18   
4096   http://sofifa.com/player/253283/facundo-pellis...        Uruguay   18   
4800   http://sofifa.com/player/251470/charles-de-ket...        Belgium   19   
5266   http://sofifa.com/player/245541/giovanni-reyna...  United States   17   
6159   http://sofifa.com/player/251852/karim-adeyemi/...        Germany   18   
6235   http://sofifa.com/player/252371/jude-bellingha...        England   17   
8116   http://sofifa.com/player/251570/rayan-

In [87]:
remove_ch = lambda x: x.replace('.', '').replace('K', '00') if isinstance(x, str) else x


In [88]:
result.Hits = result.Hits.apply(remove_ch)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\1120182670.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result.Hits = result.Hits.apply(remove_ch)


In [89]:
result.Hits.reset_index()

index  Hits
0       8  1600
1      29  1100
2      34  1500
3      68  3200
4     105  1700
5     106  2300
6     147  1900
7     416   200
8     544  1100
9     687  1300
10    690   300
11   1011  1400
12   1265  2800
13   1380  1800
14   1818  1300
15   2579  8400
16   2757  1100
17   3791  4300
18   3893  1600
19   4091  1200
20   4096  1300
21   4800  2300
22   5266  1800
23   6159  1100
24   6235  1400
25   8116   100
26   8401  2500
27  15819   600

### Some of the thousands came out to be hundred so we change those hundreds to thousands

In [90]:
index_list = [416,690,8116,15819]

In [91]:
rows_to_change = result.loc[index_list]

In [92]:
rows_to_change['Hits'] = rows_to_change['Hits'].apply(lambda x: x.replace('00', '000'))


### After updating the hundreds to thousands, update the output to the result variable which hold values that were formerly 'k's

In [93]:
result.update(rows_to_change)


C:\Users\user\AppData\Local\Temp\ipykernel_1608\1119982295.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result.update(rows_to_change)
C:\Users\user\AppData\Local\Temp\ipykernel_1608\1119982295.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result.update(rows_to_change)
C:\Users\user\AppData\Local\Temp\ipykernel_1608\1119982295.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats 

In [94]:
result

ID             Name                       FullName  \
8      231747.0        K. Mbappé                  Kylian Mbappé   
29     233049.0        J. Sancho                   Jadon Sancho   
34     212198.0  Bruno Fernandes  Bruno Miguel Borges Fernandes   
68     231677.0      M. Rashford                Marcus Rashford   
105    239085.0       E. Haaland                 Erling Haaland   
106    212462.0      Alex Telles            Alex Nicolao Telles   
147    229880.0   A. Wan-Bissaka              Aaron Wan-Bissaka   
416    220651.0         Angeliño     José Ángel Esmoris Tasende   
544    229558.0     D. Upamecano                Dayot Upamecano   
687    255475.0           Antony      Antony Matheus dos Santos   
690    244778.0          Trincão  Francisco Mota Castro Trincão   
1011   246147.0     M. Greenwood                Mason Greenwood   
1265   253004.0        Ansu Fati                 Anssumane Fati   
1380   248243.0     E. Camavinga              Eduardo Camavinga   
1818   251804.0          S. Dest                   Sergiño Dest   
2579   231485.0      A. Tuanzebe                  Axel Tuanzebe   
2757   245371.0        T. Almada                  Thiago Almada   
3791   251854.0            Pedri           Pedro González López   
3893   205343.0    J. Akpa Akpro         Jean-Daniel Akpa Akpro   
4091   254796.0       N. Madueke                   Noni Madueke   
4096   253283.0     F. Pellistri              Facundo Pellistri   
4800   251470.0  C. De Ketelaere           Charles De Ketelaere   
5266   245541.0         G. Reyna                 Giovanni Reyna   
6159   251852.0       K. Adeyemi                  Karim Adeyemi   
6235   252371.0    J. Bellingham                Jude Bellingham   
8116   251570.0        R. Cherki                   Rayan Cherki   
8401   238071.0      D. Sterling                 Dujon Sterling   
15819  255611.0       D. Maldini                 Daniel Maldini   

                                               PlayerUrl    Nationality   Age  \
8      http://sofifa.com/player/231747/kylian-mbappe/...         France  21.0   
29     http://sofifa.com/player/233049/jadon-sancho/2...        England  20.0   
34     http://sofifa.com/player/212198/bruno-miguel-b...       Portugal  25.0   
68     http://sofifa.com/player/231677/marcus-rashfor...        England  22.0   
105    http://sofifa.com/player/239085/erling-haaland...         Norway  19.0   
106    http://sofifa.com/player/212462/alex-nicolao-t...         Brazil  27.0   
147    http://sofifa.com/player/229880/aaron-wan-biss...        England  22.0   
416    http://sofifa.com/player/220651/jose-angel-esm...          Spain  23.0   
544    http://sofifa.com/player/229558/dayot-upamecan...         France  21.0   
687    http://sofifa.com/player/255475/antony-matheus...         Brazil  20.0   
690    http://sofifa.com/player/244778/francisco-mota...       Portugal  20.0   
1011   http://sofifa.com/player/246147/mason-greenwoo...        England  18.0   
1265   http://sofifa.com/player/253004/anssumane-fati...          Spain  17.0   
1380   http://sofifa.com/player/248243/eduardo-camavi...         France  17.0   
1818   http://sofifa.com/player/251804/sergino-dest/2...  United States  19.0   
2579   http://sofifa.com/player/231485/axel-tuanzebe/...        England  22.0   
2757   http://sofifa.com/player/245371/thiago-almada/...      Argentina  19.0   
3791   http://sofifa.com/player/251854/pedro-gonzalez...          Spain  17.0   
3893   http://sofifa.com/player/205343/jean-daniel-ak...    Ivory Coast  27.0   
4091   http://sofifa.com/player/254796/noni-madueke/2...        England  18.0   
4096   http://sofifa.com/player/253283/facundo-pellis...        Uruguay  18.0   
4800   http://sofifa.com/player/251470/charles-de-ket...        Belgium  19.0   
5266   http://sofifa.com/player/245541/giovanni-reyna...  United States  17.0   
6159   http://sofifa.com/player/251852/karim-adeyemi/...        Germany  18.0   
6235   http://sofifa.com/player/252371/jude-

In [95]:
df.update(result)


In [96]:
df

ID               Name                      FullName  \
0      158023.0           L. Messi                  Lionel Messi   
1       20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389.0           J. Oblak                     Jan Oblak   
3      192985.0       K. De Bruyne               Kevin De Bruyne   
4      190871.0          Neymar Jr    Neymar da Silva Santos Jr.   
...         ...                ...                           ...   
18974  247223.0             Xia Ao                        Ao Xia   
18975  258760.0           B. Hough                     Ben Hough   
18976  252757.0        R. McKinley                Ronan McKinley   
18977  243790.0       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520.0          Zhou Xiao                     Xiao Zhou   

                                               PlayerUrl Nationality   Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina  33.0   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal  35.0   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia  27.0   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium  29.0   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil  28.0   
...                                                  ...         ...   ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR  21.0   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England  17.0   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England  18.0   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR  20.0   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR  21.0   

       OverallRating   POT                 Club     Contract   Positions  \
0               93.0  93.0         FC Barcelona  2004 - 2021  RW| ST| CF   
1               92.0  92.0             Juventus  2018 - 2022      ST| LW   
2               91.0  93.0      Atlético Madrid  2014 - 2023          GK   
3               91.0  91.0      Manchester City  2015 - 2023     CAM| CM   
4               91.0  91.0  Paris Saint-Germain  2017 - 2022     LW| CAM   
...              ...   ...                  ...          ...         ...   
18974           47.0  55.0           Wuhan Zall  2018 - 2022          CB   
18975           47.0  67.0      Oldham Athletic  2020 - 2021          CM   
18976           47.0  65.0           Derry City  2019 - 2020          CM   
18977           47.0  57.0     Dalian YiFang FC  2020 - 2022          RW   
18978           47.0  57.0     Dalian YiFang FC  2019 - 2023      CB| LB   

      Height Weight Preferred Foot   BOV Best Position        Joined    Value  \
0      170cm   72kg           Left  93.0            RW   Jul 1, 2004  €103.5M   
1      187cm   83kg          Right  92.0            ST  Jul 10, 2018     €63M   
2      188cm   87kg          Right  91.0            GK  Jul 16, 2014    €120M   
3      181cm   70kg          Right  91.0           CAM  Aug 30, 2015    €129M   
4      175cm   68kg          Right  91.0            LW   Aug 3, 2017    €132M   
...      ...    ...            ...   ...           ...           ...      ...   
18974  178cm   66kg          Right  49.0            CB  Jul 13, 2018    €100K   
18975  175cm   65kg          Right  51.0           CAM   Aug 1, 2020    €130K   
18976  179cm   74kg          Right  49.0           CAM   Mar 8, 2019    €120K   
18977  175cm   69kg          Right  48.0            ST  Sep 22, 2020    €100K   
18978  188cm   75kg           Left  50.0            LB  Jul 29, 2019    €100K   

        Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0      €560K        €138.4M      429.0      85.0       95.0              70.0   
1      €220K         €75.9M      437.0      84.0       95.0              90.0   
2      €125K        €159.4M       95.0      13.0       11.0              15.0   
3      €370K          €161M      407.0      94.0 